app/scripts/lib/mnemonic.ts

```diff
@@ -0,0 +1,418 @@
+  const buffer = new ArrayBuffer(totalBytes);
+  const trieNodes = new Uint32Array(buffer, 0, totalNodes * 3);
+  const wordEndNodes = new Uint16Array(
+    buffer,
+    totalBytesForNodeArray,
+    wordList.length,
+  );
+
+  // Assign node IDs and populate arrays via BFS
+  type BFSEntry = { node: TempNode; id: number };
+  const bfsQueue: BFSEntry[] = [{ node: root, id: 0 }];
+  let nextId = 1;
+
+  while (bfsQueue.length > 0) {
+    const { node, id } = bfsQueue.shift() as BFSEntry;
+    const baseOffset = id * 3;
+
+    // Store terminalIndex and parentId
+    trieNodes[baseOffset] = packTermAndParent(
+      node.terminalIndex,
+      node.parentId,
+    );
+    if (node.terminalIndex !== NOT_TERMINAL) {
+      wordEndNodes[node.terminalIndex] = id;
+    }
+
+    // Build and store childMask and letterFromParent
+    let childMask = 0;
+    let childCount = 0;
+    for (let c = 0; c < 26; c++) {
+      if (node.children[c]) {
+        // there is another word with this same prefix, so we need to store it
+        // in the childMask.
+        childMask |= 1 << c;
+        childCount++;
+      }
+    }
+    trieNodes[baseOffset + 1] = packLetterAndMask(
+      node.letterFromParent,
+      childMask,
+    );
+
+    // Set childBase and enqueue children
+    const childBase = childCount > 0 ? nextId : 0;
+    trieNodes[baseOffset + 2] = childBase;
+    if (childCount > 0) {
+      let assigned = 0;
+      for (let c = 0; c < 26; c++) {
+        const child = node.children[c];
+        if (child) {
+          child.parentId = id;
+          child.letterFromParent = c;
+          bfsQueue.push({ node: child, id: childBase + assigned });
+          assigned++;
+        }
+      }
+      nextId += childCount;
+    }
+  }
+
+  return { trieNodes, wordEndNodes };
+}
+
+/**
+ * Finds the wordlist index of a word given as a UTF-8 byte array.
+ *
+ * @param trieNodes - trie node data
+ * @param word - UTF-8 byte array (e.g., "hello" as bytes)
+ * @returns wordlist index
+ * @throws If the word is not in the trie or contains invalid characters.
+ */
+function findWordIndex(trieNodes: Uint32Array, word: Uint8Array): number {
+  let nodeId = 0; // Start at root
+  for (const char of word) {
+    const charCode = char - 0x61;
+    if (charCode < 0 || charCode >= 26) {
+      throw new Error(ERROR_MESSAGE);
+    }
+    const baseOffset = nodeId * 3;
+    const childMask = unpackChildMask(trieNodes[baseOffset + 1]);
+    // check if the next character is in the trie at this node
+    // the childMask is a 26-bit integer where each bit represents the presence
+    // of a child node for a given character. The bit at position N is set if
+    // the child node for the character with ASCII code N+97 is present.
+    // If it is not present, the word is not in the trie.
+    if (!(childMask & (1 << charCode))) {
+      throw new Error(ERROR_MESSAGE);
+    }
+    const rank = countSetBits(childMask & ((1 << charCode) - 1));
```

Comment:
This is rank/select: https://blog.startifact.com/posts/succinct/#rank-select-bit-vector